src/envoy/http/authn/http_filter.cc

```diff
@@ -14,75 +14,66 @@
  */
 
 #include "src/envoy/http/authn/http_filter.h"
-#include "src/envoy/http/authn/mtls_authentication.h"
+#include "common/http/utility.h"
+#include "src/envoy/http/authn/origin_authenticator.h"
+#include "src/envoy/http/authn/peer_authenticator.h"
 #include "src/envoy/utils/utils.h"
 
+namespace iaapi = istio::authentication::v1alpha1;
+
 namespace Envoy {
 namespace Http {
 
 AuthenticationFilter::AuthenticationFilter(
-    const istio::authentication::v1alpha1::Policy& config)
-    : config_(config) {}
+    const istio::authentication::v1alpha1::Policy& policy)
+    : policy_(policy) {}
 
 AuthenticationFilter::~AuthenticationFilter() {}
 
 void AuthenticationFilter::onDestroy() {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
 }
 
-FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap&, bool) {
+FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
+                                                        bool) {
   ENVOY_LOG(debug, "Called AuthenticationFilter : {}", __func__);
+  state_ = Istio::AuthN::State::PROCESSING;
+
+  setHeaders(&headers);
+
+  authenticator_.reset(createPeerAuthenticator(
+      this, [this](bool success) { onPeerAuthenticationDone(success); }));
+  authenticator_->run();
```

Comment:
can these authenticator be chained into a list?   your onDone() just jump to the next one.

Reply:
The second authenticator require the result from the first to run. I find it cleaner to construct it when needed with all require data (i.e via constructor), rather than putting them in a chain and need to set some parameters accordingly.